File: src/bxct.m
% bxct Box-counting of binary image for computation of fractal dimension
% Copyright (c) 2016 Cai Jiashen

function [fd, bcFig] = bxct(I)

%Padding image to have equal dimensions
s = regionprops(I, 'BoundingBox');

if (s(1).BoundingBox(3)*s(1).BoundingBox(4) > 0.25*size(I,1)*size(I,2))
    I = imcrop(I, s(1).BoundingBox);
end


[height, width]=size(I);

if height > width
    padWidth = height-width;
    padHeight = 0;
    dimen = height;
else
    padHeight = width-height;
    padWidth = 0;
    dimen = width;
end

padI = padarray(I, [mod(padHeight,2), mod(padWidth,2)],'pre');
padI = padarray(padI,[floor(padHeight/2), floor(padWidth/2)]);

%imshow (padI);

tic;
[nBoxA, boxSizeA] = initGrid(padI,dimen);
[nBoxB, boxSizeB] = initGrid(flip(padI,1),dimen);
[nBoxC, boxSizeC] = initGrid(flip(padI,2),dimen);
[nBoxD, boxSizeD] = initGrid(flip(flip(padI,1),2),dimen);

toc

nBox = min([nBoxA; nBoxB; nBoxC; nBoxD]); % Maximal efficient covering
boxSize = boxSizeA;

totalBoxSizes = numel(boxSize);

p = polyfit(log(boxSize),log(nBox),1);
fd = -p(1);

% Box Count Plot
bcFig = figure('Name','BoxCount Plot','NumberTitle','off','Visible','off');
pause(0.01);
set(0, 'CurrentFigure', bcFig);
loglog(boxSizeA,nBoxA,'s-');
xlabel('r, box size (pixels)'); ylabel('n(r), box count');


x1=linspace(2,0.45*dimen);
y1=exp(polyval(p,log(x1)));
hold on;
loglog(x1,y1);

legend('Box Count','Regression');

    function [nBox, boxSize]= initGrid(I,dimen)
        
        startBoxSize = floor(0.45*dimen);
        curBoxSize = startBoxSize;
        
        nBox = zeros(1,(startBoxSize-2+1));
        boxSize= [startBoxSize:-1:2];
        for sizeCount = 1:(startBoxSize-2+1)
            curBoxSize = boxSize(sizeCount);
            
            for macroY = 1:ceil(dimen/curBoxSize)
                for macroX = 1:ceil(dimen/curBoxSize)
                    boxYinit = (macroY-1)*curBoxSize+1;
                    boxXinit = (macroX-1)*curBoxSize+1;
                    boxYend = min(macroY*curBoxSize,dimen);
                    boxXend = min(macroX*curBoxSize,dimen);
                    
                    boxFound = false;
                    for curY = boxYinit:boxYend
                        for curX = boxXinit:boxXend
                            if I(curY,curX)
                                boxFound = true;
                                nBox(sizeCount) = nBox(sizeCount) + 1;
                                break;
                            end
                        end
                        
                        if boxFound == true
                            break;
                        end
                    end
                    
%                     if ~isempty(find(I(boxYinit:boxYend,boxXinit:boxXend),1,'first'))
%                         nBox(sizeCount) = nBox(sizeCount)+1;
%                     end
                    
%                     if sum(sum(I(boxYinit:boxYend,boxXinit:boxXend))) > 0
%                         nBox(sizeCount) = nBox(sizeCount)+1;
%                     end
                    
                end
            end
        end    
    end
end

File: python/src/bxct.py
import numpy as np
import matplotlib.pyplot as plt
from skimage.measure import regionprops
from skimage.util import pad

def bxct(I):
    # Padding image to have equal dimensions
    props = regionprops(I)
    bbox = props[0].bbox

    if bbox[2] * bbox[3] > 0.25 * I.shape[0] * I.shape[1]:
        I = I[bbox[0]:bbox[2], bbox[1]:bbox[3]]

    height, width = I.shape
    if height > width:
        pad_width = height - width
        pad_height = 0
        dimen = height
    else:
        pad_height = width - height
        pad_width = 0
        dimen = width

    padI = pad(I, ((pad_height % 2, pad_height // 2), (pad_width % 2, pad_width // 2)), mode='constant')

    # Initialize grid and perform box-counting
    nBoxA, boxSizeA = init_grid(padI, dimen)
    nBoxB, boxSizeB = init_grid(np.flip(padI, axis=0), dimen)
    nBoxC, boxSizeC = init_grid(np.flip(padI, axis=1), dimen)
    nBoxD, boxSizeD = init_grid(np.flip(np.flip(padI, axis=0), axis=1), dimen)

    nBox = np.minimum.reduce([nBoxA, nBoxB, nBoxC, nBoxD])
    boxSize = boxSizeA

    # Linear regression to find fractal dimension
    p = np.polyfit(np.log(boxSize), np.log(nBox), 1)
    fd = -p[0]

    # Box Count Plot
    plt.figure()
    plt.loglog(boxSizeA, nBoxA, 's-')
    x1 = np.linspace(2, 0.45 * dimen, num=100)
    y1 = np.exp(np.polyval(p, np.log(x1)))
    plt.loglog(x1, y1)
    plt.xlabel('r, box size (pixels)')
    plt.ylabel('n(r), box count')
    plt.legend(['Box Count', 'Regression'])
    plt.show()

    return fd

def init_grid(I, dimen):
    start_box_size = int(0.45 * dimen)
    nBox = np.zeros(start_box_size - 1)
    boxSize = np.arange(start_box_size, 1, -1)
    
    for sizeCount, curBoxSize in enumerate(boxSize):
        for macroY in range(int(np.ceil(dimen / curBoxSize))):
            for macroX in range(int(np.ceil(dimen / curBoxSize))):
                boxYinit = macroY * curBoxSize
                boxXinit = macroX * curBoxSize
                boxYend = min(boxYinit + curBoxSize, dimen)
                boxXend = min(boxXinit + curBoxSize, dimen)

                boxFound = np.any(I[boxYinit:boxYend, boxXinit:boxXend])
                if boxFound:
                    nBox[sizeCount] += 1

    return nBox, boxSize
